### PR TITLE
no_APR_Step4

### PR DESCRIPTION
**Pull request by dkeeney**
- remove GraphAlgorithms altogether
- fixed Directory, Path, OS   did full APR removal.
- corresponding tests. modified 
- StringUtils so it would compile but did not do the full APR removal yet.

Part of #81 
Replaces #79

### CODE REVIEW

**Comment by dkeeney** [comment]
@breznak  here is Step4 to review

**Comment by dkeeney** [comment]
These are good comments.
I will put them in.  Thanks.

File: src/nupic/os/ImportFilesystem.hpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#else
+  // C++11
+  #include <boost/filesystem.hpp>
+  #define USE_BOOST_FILESYSTEM 1
+#endif
+
+#ifdef USE_BOOST_FILESYSTEM
+  namespace fs = boost::filesystem;
+  namespace er = boost::system;
+  #define  FS_PermNone    fs::perms::no_perms
```
Comment (breznak) [inline src/nupic/os/ImportFilesystem.hpp]: `using FS_PermNone = fs:....` 
would be better as it'd be all pure c++, no macros
...but it's fine as is, if you don't wanna change

File: src/nupic/os/Path.cpp [hunk 2]
```diff
@@ -362,28 +365,13 @@ void Path::setPermissions(const std::string &path, bool userRead,
...
-                  : fs::perms::none) |
-        (userWrite ? fs::perms::owner_all : fs::perms::none) |
-        (groupRead ? fs::perms::group_exec | fs::perms::group_read
-                   : fs::perms::none) |
-        (groupWrite ? fs::perms::group_all : fs::perms::none) |
-        (otherRead ? fs::perms::others_exec | fs::perms::others_read
-                   : fs::perms::none) |
-        (otherWrite ? fs::perms::others_all : fs::perms::none);
-#endif
+	    (userRead ? fs::perms::owner_exe | fs::perms::owner_read : FS_PermNone) |
```
Comment (breznak) [inline src/nupic/os/Path.cpp]: this looks nice :)

File: src/nupic/os/Timer.cpp [hunk 3]
```diff
@@ -27,134 +27,62 @@
...
-  }
-}
-
-static inline nupic::UInt64 getCurrentTime() {
-  return (nupic::UInt64)(mach_absolute_time() - initialTicks_);
-}
-
-static inline nupic::UInt64 getTicksPerSec() { return ticksPerSec_; }
-
-#else
```
Comment (breznak) [inline src/nupic/os/Timer.cpp]: great cleanup here!

**Comment by breznak** [comment]
Great job @dkeeney ! :clap: What's your next endeavor?

**Comment by dkeeney** [comment]
I did not like #define either but these are enum values not types or namespace so I could not find any other easy way to do it.

Next is noAPR_Step5

**Pull request merged**